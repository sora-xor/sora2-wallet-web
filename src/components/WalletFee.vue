<template>
  <div class="info-line-container">
    <info-line
      class="wallet-fee"
      is-formatted
      :label="t('networkFeeText')"
      :label-tooltip="t('networkFeeTooltipText')"
      :value="formattedValue"
      :asset-symbol="xor"
      :fiat-value="formattedFiatValue"
    />
  </div>
</template>

<script lang="ts">
import { FPNumber } from '@sora-substrate/sdk';
import { XOR } from '@sora-substrate/sdk/build/assets/consts';
import { Component, Mixins, Prop } from 'vue-property-decorator';

import InfoLine from './InfoLine.vue';
import FormattedAmountMixin from './mixins/FormattedAmountMixin';
import TranslationMixin from './mixins/TranslationMixin';

function isFPNumber(value: unknown): boolean {
  if (value instanceof FPNumber) {
    return true;
  }
  throw new Error('[WalletFee.vue]: property "value" should have FPNumber type');
}

@Component({
  components: {
    InfoLine,
  },
})
export default class WalletFee extends Mixins(TranslationMixin, FormattedAmountMixin) {
  @Prop({ required: true, validator: isFPNumber }) readonly value!: FPNumber;

  get xor(): string {
    return XOR.symbol;
  }

  private formatValue(value: FPNumber): string {
    return value.isFinity() ? value.toLocaleString() : value.toString();
  }

  get formattedValue(): string {
    return this.formatValue(this.value);
  }

  get formattedFiatValue(): Nullable<string> {
    const value = this.getFPNumberFiatAmountByFPNumber(this.value);
    if (!value) {
      return value;
    }
    return this.formatValue(value);
  }
}
</script>
