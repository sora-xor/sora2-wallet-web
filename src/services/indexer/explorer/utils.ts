import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
