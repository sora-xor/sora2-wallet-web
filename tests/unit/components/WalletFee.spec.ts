import Vuex from 'vuex'
import { shallowMount } from '@vue/test-utils'

import { useDescribe, localVue } from '../../utils'
import { MOCK_WALLET_FEE } from '../../utils/WalletFeeMock'

import WalletFee from '@/components/WalletFee.vue'

useDescribe('WalletFee.vue', WalletFee, () => {
  const store = new Vuex.Store({
    getters: {
      fiatPriceAndApyObject: () => ({
        '0x0200000000000000000000000000000000000000000000000000000000000000': {
          price: '1230000000000000000'
        }
      })
    }
  })
  MOCK_WALLET_FEE.map(item => it(`[${item.title}]: should be rendered correctly`, () => {
    const propsData = {
      value: item.value
    }
    const wrapper = shallowMount(WalletFee, { localVue, store, propsData })
    expect(wrapper.element).toMatchSnapshot()
  }))
  it('Should not be rendered', () => {
    const propsData = {
      value: '123'
    }
    try {
      shallowMount(WalletFee, { localVue, store, propsData })
    } catch (error) {
      expect((error as Error).message).toBe('[WalletFee.vue]: property "value" should have FPNumber type')
    }
  })
})
