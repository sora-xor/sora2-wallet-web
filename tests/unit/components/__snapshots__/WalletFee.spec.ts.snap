// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletFee.vue [With "0" Value]: should be rendered correctly 1`] = `
<div
  class="info-line-container"
>
  <info-line-stub
    assetsymbol="XOR"
    class="wallet-fee"
    fiatvalue="0"
    isformatted="true"
    label="Network Fee"
    labeltooltip="Network fee is used to ensure SORA system's growth and stable performance."
    value="0"
    valuetooltip=""
  />
</div>
`;

exports[`WalletFee.vue [With "Infinity" Value]: should be rendered correctly 1`] = `
<div
  class="info-line-container"
>
  <info-line-stub
    assetsymbol="XOR"
    class="wallet-fee"
    fiatvalue="Infinity"
    isformatted="true"
    label="Network Fee"
    labeltooltip="Network fee is used to ensure SORA system's growth and stable performance."
    value="Infinity"
    valuetooltip=""
  />
</div>
`;

exports[`WalletFee.vue [With "NaN" Value]: should be rendered correctly 1`] = `
<div
  class="info-line-container"
>
  <info-line-stub
    assetsymbol="XOR"
    class="wallet-fee"
    fiatvalue="NaN"
    isformatted="true"
    label="Network Fee"
    labeltooltip="Network fee is used to ensure SORA system's growth and stable performance."
    value="NaN"
    valuetooltip=""
  />
</div>
`;

exports[`WalletFee.vue [With Value]: should be rendered correctly 1`] = `
<div
  class="info-line-container"
>
  <info-line-stub
    assetsymbol="XOR"
    class="wallet-fee"
    fiatvalue="1,518,518.5048512"
    isformatted="true"
    label="Network Fee"
    labeltooltip="Network fee is used to ensure SORA system's growth and stable performance."
    value="1,234,567.890123"
    valuetooltip=""
  />
</div>
`;
